dask/dataframe/operation.py

```diff
@@ -0,0 +1,596 @@
+from __future__ import annotations
+
+import itertools
+import operator
+from collections.abc import Mapping
+from functools import singledispatch
+from typing import Any
+
+import numpy as np
+
+from dask.base import tokenize
+from dask.dataframe.utils import make_meta
+from dask.highlevelgraph import HighLevelGraph
+from dask.operation import CollectionOperation, MemoizingVisitor, operations, regenerate
+from dask.utils import apply, is_arraylike
+
+
+class DataFrameOperation(CollectionOperation):
+
+    _name: str
+    _meta: Any
+    _divisions: tuple
+    _columns: set | None = None
+    applied_filters: list | None = None
+    projected_columns: set | None = None
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @name.setter
+    def name(self, value):
+        self._name = value
+
+    @property
+    def meta(self):
+        return self._meta
+
+    @meta.setter
+    def meta(self, value):
+        self._meta = value
+
+    @property
+    def divisions(self) -> tuple:
+        return self._divisions
+
+    @divisions.setter
+    def divisions(self, value):
+        self._divisions = value
+
+    @property
+    def npartitions(self) -> int:
+        return len(self.divisions) - 1
+
+    @property
+    def collection_keys(self) -> list[tuple]:
+        return [(self.name, i) for i in range(self.npartitions)]
+
+    @property
+    def columns(self):
+        return self._columns
+
+    @classmethod
+    def optimize_collection(cls, collection):
+        if hasattr(collection, "operation") and isinstance(
+            collection.operation, DataFrameOperation
+        ):
+            return type(collection)(operation=optimize(collection.operation))
+        return collection
```

Comment:
Should an operation be responsible for optimizing a collection? Seems like the wrong level of abstraction to me. Is this used somewhere?